Convert unbatched inputs to batch with BS=1, and Make MHA symbolically traceable

mikekgfb opened the pull request:
Summary: Make MHA symbolically traceable

Test Plan: sandcastle

Differential Revision: D39790429

malfet commented:
/easycla

mikekgfb commented:
________________________________
From: Nikita Shulga ***@***.***>
Sent: Monday, October 10, 2022 4:10 PM
To: pytorch/pytorch ***@***.***>
Cc: Michael Gschwind ***@***.***>; Author ***@***.***>
Subject: Re: [pytorch/pytorch] Make MHA symbolically traceable (PR #85643)

/easycla — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you authored the thread. Message ID: <pytorch/pytorch/pull/85643/c1273898348@ github. com> ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍
ZjQcmQRYFpfptBannerStart
This Message Is From an External Sender

ZjQcmQRYFpfptBannerEnd

/easycla

—
Reply to this email directly, view it on GitHub<https://github.com/pytorch/pytorch/pull/85643#issuecomment-1273898348>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AOT4XHPKZZKOVOIBCNP2S6DWCSPAHANCNFSM6AAAAAAQWA4MIM>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

malfet commented:
/easycla

ngimel commented on torch/nn/modules/activation.py:
```diff
@@ -1061,7 +1061,19 @@ def forward(self, query: Tensor, key: Tensor, value: Tensor, key_padding_mask: O
+            key = key.view(-1, ksize[-2], ksize[-1])
+            value = value.view(-1, vsize[-2], vsize[-1])
+        else:
+            query = query.view(qsize[0], qsize[1], -1)
```
this doesn't look right, for non-batch-first sizes are (S,N,E) or (S,E), so it should be `.view(size(0), -1, size(-1)`

Commit:
make unbatched inputs batches w/ BS=1 (#85643)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/85643

Make MHA symbolically traceable

Test Plan: sandcastle

Differential Revision: D39790429

fbshipit-source-id: 4a4c30d70792a3fc22cb84f08c130b29491f0d7d
